src/main/java/com/societegenerale/commons/plugin/rules/DontReturnNullCollectionTest.java

```diff
@@ -0,0 +1,66 @@
+package com.societegenerale.commons.plugin.rules;
+
+import static com.tngtech.archunit.lang.syntax.ArchRuleDefinition.methods;
+
+import java.util.Collection;
+import java.util.List;
+import java.util.Set;
+import java.util.stream.Collectors;
+
+import javax.annotation.Nonnull;
+
+import com.societegenerale.commons.plugin.utils.ArchUtils;
+import com.tngtech.archunit.base.DescribedPredicate;
+import com.tngtech.archunit.core.domain.JavaClass;
+import com.tngtech.archunit.core.domain.JavaMethod;
+import com.tngtech.archunit.core.importer.ClassFileImporter;
+import com.tngtech.archunit.lang.ArchRule;
+
+/**
+ * Returning null collections (List, Set) forces the caller to always perform a
+ * null check, which hinders readability. It's much better to never return a
+ * null Collection, and instead return an empty one. This rule enforces that all
+ * methods returning a Collection must be annotated with @Nonnull
+ *
+ * @see: there is no agreed standard for notNull annotation, see <a href=
+ *       "https://stackoverflow.com/questions/4963300/which-notnull-java-annotation-should-i-use/">here</a>
+ *
+ */
+public class DontReturnNullCollectionTest implements ArchRuleTest {
+
+	protected static final String NO_NULL_COLLECTION_MESSAGE = "we don't want callers to perform null check every time. Return an empty collection, not null.";
+
+	@Override
+	public void execute(String path) {
+
+		// Using JavaClass to get SubClasses
+
+		JavaClass javaClassCollection = new ClassFileImporter().importClass(Collection.class);
+
+		DescribedPredicate<JavaMethod> returnACollectionSubClass = new DescribedPredicate<JavaMethod>(
+				"return a Collection SubClass") {
+			@Override
+			public boolean apply(JavaMethod method) {
+
+				Set<JavaClass> javaClasscollectionSubClasses = javaClassCollection.getAllSubClasses();
+
+				// A Set containing the equivalent using reflect
+
+				Set<Class<?>> equivalentToJavaClasscollectionSubClasses = javaClasscollectionSubClasses.parallelStream()
+						.map(JavaClass::reflect).collect(Collectors.toSet());
+
+				return equivalentToJavaClasscollectionSubClasses.contains(method.getRawReturnType().reflect());
+
+			}
+		};
+
+		ArchRule rule = methods().that().haveRawReturnType(List.class).or().haveRawReturnType(Set.class).should()
+				.beAnnotatedWith(Nonnull.class).because(NO_NULL_COLLECTION_MESSAGE);
+
+		ArchRule ruleUsingCollectionSubClasses = methods().that(returnACollectionSubClass).should()
+				.beAnnotatedWith(Nonnull.class).because(NO_NULL_COLLECTION_MESSAGE);
```

Comment:
@FanJups I don't know, if I correctly understand what you're trying to do, but is there any reason not to use

```
methods().that().haveRawReturnType(JavaClass.Predicates.assignableTo(Collection.class))
```

? `reflect()` is always dangerous, because it relies on the classpath. Everything else can be asserted on any `.class` file anywhere in the file system...

Reply:
@codecholeric, it looks it doesn't work.

Reply:
@codecholeric, I've tried this option but one test failed. 
`Running com.societegenerale.commons.plugin.rules.DontReturnNullCollectionTestTest
Tests run: 2, Failures: 1, Errors: 0, Skipped: 0, Time elapsed: 2.817 sec <<< FAILURE!
shouldThrowViolations(com.societegenerale.commons.plugin.rules.DontReturnNullCollectionTestTest)  Time elapsed: 1.91 sec  <<< FAILURE!
java.lang.AssertionError: 
Expecting actual not to be null
	at com.societegenerale.commons.plugin.rules.DontReturnNullCollectionTestTest.shouldThrowViolations(DontReturnNullCollectionTestTest.java:35)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:497)
	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:50)
	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:47)
	at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
	at org.junit.internal.runners.statements.RunBefores.evaluate(RunBefores.java:26)
	at org.junit.runners.ParentRunner.runLeaf(ParentRunner.java:325)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:78)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:57)
	at org.junit.runners.ParentRunner$3.run(ParentRunner.java:290)
	at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:71)
	at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:288)
	at org.junit.runners.ParentRunner.access$000(ParentRunner.java:58)
	at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:268)
	at org.junit.runners.ParentRunner.run(ParentRunner.java:363)
	at org.apache.maven.surefire.junit4.JUnit4Provider.execute(JUnit4Provider.java:252)
	at org.apache.maven.surefire.junit4.JUnit4Provider.executeTestSet(JUnit4Provider.java:141)
	at org.apache.maven.surefire.junit4.JUnit4Provider.invoke(JUnit4Provider.java:112)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:497)
	at org.apache.maven.surefire.util.ReflectionUtils.invokeMethodWithArray(ReflectionUtils.java:189)
	at org.apache.maven.surefire.booter.ProviderFactory$ProviderProxy.invoke(ProviderFactory.java:165)
	at org.apache.maven.surefire.booter.ProviderFactory.invokeProvider(ProviderFactory.java:85)
	at org.apache.maven.surefire.booter.ForkedBooter.runSuitesInProcess(ForkedBooter.java:115)
	at org.apache.maven.surefire.booter.ForkedBooter.main(ForkedBooter.java:75)
`

`package com.societegenerale.commons.plugin.rules;

import static org.assertj.core.api.Assertions.assertThat;
import static org.assertj.core.api.Assertions.assertThatCode;
import static org.assertj.core.api.Assertions.catchThrowable;

import org.junit.Before;
import org.junit.Test;

import com.societegenerale.commons.plugin.SilentLog;
import com.societegenerale.commons.plugin.utils.ArchUtils;

public class DontReturnNullCollectionTestTest {

	String pathObjectWithAmethodReturningAnullList = "./target/aut-target/classes/com/societegenerale/aut/main/ObjectWithMethodsReturningNullCollections.class";

	String pathProperlyAnnotatedObjectWithAmethodReturningAlist = "./target/aut-target/classes/com/societegenerale/aut/main/ObjectWithProperlyAnnotatedMethodsReturningCollections.class";

	@Before
	public void setup() {
		// in the normal lifecycle, ArchUtils is instantiated, which enables a static
		// field there to be initialized
		ArchUtils archUtils = new ArchUtils(new SilentLog());
	}

	@Test
	public void shouldThrowViolations() {

		Throwable validationExceptionThrown = catchThrowable(() -> {

			new DontReturnNullCollectionTest().execute(pathObjectWithAmethodReturningAnullList);

		});

		assertThat(validationExceptionThrown).isInstanceOf(AssertionError.class)
				.hasMessageContaining("was violated (2 times)");

	}

	@Test
	public void shouldNotThrowViolations() {

		assertThatCode(
				() -> new DontReturnNullCollectionTest().execute(pathProperlyAnnotatedObjectWithAmethodReturningAlist))
						.doesNotThrowAnyException();

	}
}`

`package com.societegenerale.aut.main;

import java.util.List;
import java.util.Set;

public class ObjectWithMethodsReturningNullCollections {
	public List returningANullList() {
		return null;
	}

	public Set returningANullSet() {
		return null;
	}

}
`

My goal was to get all the subclasses of Collection then verify that each method returns a type present in **equivalentToJavaClasscollectionSubClasses** object.

I used **reflect** because I wanted to get the original class equivalent to its JavaClass object.

Just want to apologize for the inconvenience. I resolved the conversation because the PR is already closed.